### PR TITLE
Issues/374

### LINKED ISSUE #367: Let users specify 'jboss' or 'rhel' along with other facts
SImple change to scancommand.py

### LINKED ISSUE #374: Make fact definition more robust
We could make it impossible to cause #373 again by having a fact-defining function that automatically adds a fact to the right lists. I.e. instead of
```
MY_TUPLE = ('my-new-fact')
...
OTHER_TUPLE = ... + MY_TUPLE
```

it should be

```
new_fact('my-new-fact', categories=[SENSITIVE_FACTS, GREEN_FACTS])
```

`new_fact` would add facts to `ALL_FACTS` automatically.

### PR DESCRIPTION
**Pull request by noahl**
Introduce a function `new_fact` which defines a new fact and
automatically adds it to the list of all facts. This makes it
impossible to accidentally leave it out, preventing bug #373 from
happening again.

Also make `new_fact` function be the source of fact documentation, and
require documentation for every new fact, which will prevent bugs like

Finally, also make fact groups be inclusive rather than exclusive, to
match customer expectations.

Closes #374 . Closes #367 .

### CODE REVIEW

File: rho/facts.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            facts_to_collect.update(RHEL_FACTS)
+        elif fact == 'jboss':
+            facts_to_collect.update(JBOSS_FACTS)
+        else:
+            facts_to_collect.add(fact)
+
+    facts_to_collect.update(ALWAYS_COLLECT)
+
+    if not facts_to_collect.issubset(ALL_FACTS):
+        log.error('Bad fact list %s', facts_to_collect)
```
Comment (chambridge) [inline rho/facts.py]: It would be nicer here if we output the difference that shows the facts that are actually invalid in the subset.

**Review by chambridge — COMMENTED** [review]
Looks good.

Need to update the Contributing.rst with these instructions:
https://github.com/quipucords/rho/blob/master/CONTRIBUTING.rst#adding-facts

**Review by elyezer — APPROVED** [review]
I really like these changes as it deduplicate a lot of work needed to get fact information across different rho subcommands and in the documentation.

Please wait for at least one more reviewer as this is a change that touches many files.

I will run Camayoc tests with this patch to make sure it continue returning the expected fact list.

File: rho/facts.py (hunk 1 above)
Comment (noahl) [reply rho/facts.py]: Done, thanks! Sample output:
```
> PYTHONPATH=rho:$PYTHONPATH bin/rho scan --profile=jboss-test-server --facts jboss.eap.installed-versions --facts jboss.eap.deploy-dates --reportfile=out.csv -vvv --facts fizzbuzz --facts abc
Unknown facts requested: abc, fizzbuzz
```

**Comment by noahl** [comment]
@chambridge I just uploaded a new version that updates CONTRIBUTING.rst and prints the list of bad facts when we have it. Please take a look.

**Review by chambridge — APPROVED** [review]

**Pull request merged**